dbmanager/src/dbmanager.c

```diff
@@ -21,18 +22,131 @@ struct DB {
   dbNode* head;
 };
 
+
+/*  function to create a new entry for the database.
+ *  @key - a unique integer value.
+ *  @val - pointer to the data stored.
+ *  @return - pointer if success, NULL if failure.
+ */
+static dbNode* create_node(int key, void* val) {
+  dbNode* newNode = (dbNode*)malloc(sizeof(dbNode));
+  return newNode;
+}
+
+
 /*
- *  keeps track of head of database.
+ *  deletes all the nodes from the given tree.
+ *  @head - head of the tree to be deleted.
  */
-dbNode* dbHead = NULL;
+static void delete_tree(dbNode* head) {
 
-/*  function to create a new entry for the database.
+}
+
+
+/*
+ *  inserts node into the given tree.
+ *  @dbCxt - database context to operate on.
+ *  @node - node to be inserted.
+ */
+static void insert_node(DB* dbCxt, dbNode* node) {
+   
+}
+
+
+/*
+ *  deletes node from the given tree.
+ *  @dbCxt - database context to operate on.
+ *  @key - key to be deleted.
+ *  @return - 0 on success, -1 on failure to delete.
+ */
+static int delete_node(DB* dbCxt, int key) {
+  return 0;
+}
+
+
+/*
+ *  returns value corresponding to provided key from the given tree.
+ *  @dbCxt - database context to operate on.
+ *  @key - key to be searched for.
+ *  @return - value pointer on success, NULL on failure to find key.
+ */
+static void* get_val(DB* dbCxt, int key) {
+  return NULL;
+}
+
+
+/*
+ *  initializes database.
+ *  can only be called once for single database application.
+ *  @return - pointer to database context if success, NULL if failure.
+ */
+DB* db_init() {
+  DB* newDb = (DB*)malloc(sizeof(DB));
+  printf("dbManager initialized!!\n");
+  return newDb;
+}
+
+
+/*
+ *  destroys database.
+ *  @dataBase - context of the database to be operated on.
+ *  @return - 0 if success, -1 if failure.
+ */
+int db_destroy(DB* dataBase) {
+  if(!dataBase) return -1;
```

Comment:
There can always be an argument about how much error checking one can do when looking at the argument list. You can take a stance and say that if they pass in the wrong data its a breach of the modules contract and you dont do any error checking for arguments.

I leave it up to you but in in general there is limited chance of the context being wrong (memory corruption) so there is not much gain from just checking this. Also there is a larger risk which is that the context is being referenced after is been deleted (common mistake in programming, use of freed memory). To combat this more likely scenario you have a few options;

1. Add a magic int field and set it to specific values for active/deleted scenarios
2. Dont return the pointer to the struct directly, instead return a pointer to a pointer to a struct. In this scenario you can set the ptr to null when the context is destroyed and then check for null pointer to catch the use after free scenario.

Option 2 is actually just a more complex form of option 1 in reality, also you can add a DB* entry to the struct itself to keep it as one malloc.